hcipy/optics/segmented_mirror.py

```diff
@@ -0,0 +1,237 @@
+        self.segnum = len(seg_pos.x)
+        self.segmentlist = np.arange(1, self.segnum + 1)
+        self._coef = np.zeros((self.segnum, 3))
+        self.seg_pos = seg_pos
+        self.input_grid = aperture.grid
+        self._last_npix = np.nan  # see _setup_grids for this
+
+    def forward(self, wavefront):
+        """Propagate a wavefront through the segmented mirror.
+
+        Parameters
+        ----------
+        wavefront : Wavefront
+            The incoming wavefront.
+
+        Returns
+        -------
+        Wavefront
+            The reflected wavefront.
+        """
+        wf = wavefront.copy()
+        wf.electric_field *= np.exp(2j * self.surface * wavefront.wavenumber)
+        return wf
+
+    def backward(self, wavefront):
+        """Propagate a wavefront backwards through the deformable mirror.
+
+        Parameters
+        ----------
+        wavefront : Wavefront
+            The incoming wavefront.
+
+        Returns
+        -------
+        Wavefront
+            The reflected wavefront.
+        """
+        wf = wavefront.copy()
+        wf.electric_field *= np.exp(-2j * self.surface * wavefront.wavenumber)
+        return wf
+
+    @property
+    def surface(self):
+        """ The surface of the segmented mirror in meters, the full surface as a Field.
+        """
+        surf = self.apply_coef()
+        return surf
+
+    @property
+    def coef(self):
+        """ The surface shape of the deformable mirror, in meters and radians; PTT segment coefficients.
+        """
+        return self._coef
+
+    def flatten(self):
+        """ Flatten the DM by setting all segment coefficients to zero."""
+        self._coef[:] = 0
+
+    def set_segment(self, segid, piston, tip, tilt):
+        """ Set an individual segment of the DM.
+
+        Piston in meter of surface, tip and tilt in radians of surface.
+
+        Parameters
+        -------------
+        segid : integer
+            Index of the segment you wish to control, starting at 1 (center would  be 0, but doesn't exist)
+        piston, tip, tilt : floats, meters and radians
+            Piston (in meters) and tip and tilt (in radians)
+        """
+        self._coef[segid - 1] = [piston, tip, tilt]
+
+    def _setup_grids(self):
+        """ Set up the grids to compute the segmented mirror surface into.
+        This is relatively slow, but we only need to do this once for
+        each size of input grids.
+        """
+        npix = self.aperture.shaped.shape[0]
+        if npix == self._last_npix:
+            return
+        else:
+            self._last_npix = npix
+
+        x, y = self.input_grid.coords
+
+        self._seg_mask = np.zeros_like(x)
+        self._seg_x = np.zeros_like(x)
+        self._seg_y = np.zeros_like(y)
+        self._seg_indices = dict()
+
+        pupil_grid = make_pupil_grid(dims=npix, diameter=PUP_DIAMETER)
+        aper_num, seg_positions = get_atlast_aperture(normalized=False,
```

Comment:
It doesn't seem like a good solution to evaluate the aperture inside the class, even if it works for now, because 1) I already pass in the aperture generator and the segment positions when I instantiate the class and 2) it makes the segmented mirror only work for one specific aperture, which is very restrictive. This definitely needs to be changed.

Reply:
You can pass in the `aper_num` and `segment_positions` as arguments to the class. You can then also get the pupil grid from that `aper_num`. Also rename `aper_num` to `subaperture_indices`.

Reply:
This in-situ aperture evaluation is here because the segment transmissions used here provides the segment identification for the mirror. I fell like that could also be something returned by the aperture directly, but I’ve realized we’ve started talking about adding a lot of function returns for the segmented apertures to accommodate the `SegmentedMirror()` functionalities. Should we just proceed with this? Can you think of a more elegant solution for passing in the segment numbering? Would it make sense of adding a class that uses `Field` as its base class which we can use to pass apertures around that are already evaluated on a grid, and that have attributes like its segment positions, diameter, segment numbering etc.?

Reply:
Alternatively, we could pass the aperture generator and a pupil grid into the `SegmentedMirror()` class, instead of the sampled aperture. This would still lead to the evaluation of the aperture in the class, but it would make it agnostic to other segmented apertures and with caching, the evaluation would happen only once or twice for an instance of the class.

Reply:
See https://github.com/ehpor/hcipy/pull/15#issuecomment-495401660